src/main/kotlin/io/libp2p/core/transport/Transport.kt

```diff
@@ -0,0 +1,47 @@
+package io.libp2p.core.transport
+
+import io.libp2p.core.Connection
+import io.libp2p.core.multiformats.Multiaddr
+import java.util.concurrent.CompletableFuture
+
+/**
+ * A Transport represents an adapter to integrate (typically) a L2, L3 or L7 protocol into libp2p, to allow incoming
+ * and outgoing connections to be established. A Transport
+ *
+ * TODO:
+ *   * Expose transport qualities/attributes (dial only, listen only, reliable, connectionful, costly, etc.)
+ */
+interface Transport {
+
+    /**
+     * Verifies whether this transport is capable of handling this multiaddr.
+     */
+    fun handles(addr: Multiaddr): Boolean
+
+    /**
+     * Performs any preparation, warm-up or initialisation to activate this transport.
+     */
+    fun initialize()
+
+    /**
+     * Stops the transport entirely, closing down all ongoing connections, outbound dials, and listening endpoints.
+     */
+    fun close(): CompletableFuture<Void>
+
+    /**
+     * Makes this transport listen on this multiaddr. The future completes once the endpoint is effectively listening.
+     */
+    fun listen(addr: Multiaddr): CompletableFuture<Void>
```

Comment:
How incoming connections are handled? It would be perfect to return here something like
```kt
Publisher<Connection>  // RX style

or

Channel<Connection>  // Kotlin style
```

Reply:
Or with the Netty style it could be 
```kt
fun listen(addr: Multiaddr, chInit: ChannelInitializer): CompletableFuture<Void>
```

Reply:
So I don't think we'd return a specific connection here. This method begins _listening_ on a particular multiaddr. In Netty terms, it probably starts a `ServerBootstrap`. We need to attach a handler to the server to process `channelActive` events (incoming sockets) and upgrade them (negotiate the secure channel and multiplexing).

At that point the `Transport` will have a `Connection` and it'll need to register it in the `Network` by calling `Network#addConnection()`, so we need to pass a reference to the `Network` into the `Transport` (maybe in the constructor for now?).

We also need to be able to remove connections when the `channelInactive` event fires.

Note that protocols and user-land code can subscribe to `Connected` and `Disconnected` notifications at the libp2p layer. See a notifee in action here: https://github.com/libp2p/go-libp2p-kad-dht/blob/master/notif.go, and the way it registers itself with the Network: https://github.com/libp2p/go-libp2p-kad-dht/blob/master/dht.go#L90

We should probably NOT use this exact approach, but rather the local event bus, which we landed recently in go-libp2p: https://github.com/libp2p/go-libp2p/issues/653.